examples/network_traffic_prediction/nr_models.py

```diff
@@ -0,0 +1,702 @@
+        """
+
+        # ====== Not do evaluation in the auto-agressive inference ====== #
+
+        # ====== Not save embedding in this customized inferrer ========= #
+        assert save_embed_path is None, \
+            "Unable to save the node embeddings when using mini batch inference." \
+            "It is not guaranteed that mini-batch prediction will cover all the nodes."
+
+        self._model.eval()
+
+        # get the graph
+        g = loader.data.g
+        # get the model metadata
+        window_size = self._model.window_size
+
+        # This customized inferre only handle the first ntypes for inference.
+        infer_ntype = list(loader.target_nidx.keys())[0]
+
+        # get all demands features
+        inventory_nfeats = g.nodes[infer_ntype].data['inventory_amounts']
+        th_inventory_nfeats = inventory_nfeats[th.arange(inventory_nfeats.shape[0])]
+
+        predict_results = None
+        for predict_step in range(init_step, (init_step + infer_steps)):
+            res = node_mini_batch_gnn_predict(self._model, loader, return_proba,
+                                                return_label=False, use_ar=True,
+                                                predict_step=predict_step)
+            preds = res[0]
+
+            replace_step = predict_step + window_size
+            # Replace the next step inventory_amounts with the predicitons
+            th_inventory_nfeats[:, replace_step] = preds[infer_ntype].squeeze()
+            inventory_nfeats[th.arange(th_inventory_nfeats.shape[0])] = th_inventory_nfeats
+            g.nodes[infer_ntype].data['inventory_amounts'] = inventory_nfeats
+
+            if predict_results is None:
+                predict_results = preds[infer_ntype]
+            else:
+                predict_results = th.concat([predict_results, preds[infer_ntype]], dim=-1)
+
+        total_predict_results = {infer_ntype: predict_results}
+
+        nid_shuffler = None
+        g = loader.data.g
+
+        if save_prediction_path is not None:
+            # save_embed_path may be None. In that case, we need to init nid_shuffler
+            if nid_shuffler is None:
+                nid_shuffler = NodeIDShuffler(g, node_id_mapping_file,
+                                              list(total_predict_results.keys())) \
+                    if node_id_mapping_file else None
+            shuffled_preds = {}
+            for ntype, pred in total_predict_results.items():
+                pred_nids = loader.target_nidx[ntype]
+                if node_id_mapping_file is not None:
+                    pred_nids = nid_shuffler.shuffle_nids(ntype, pred_nids)
+                shuffled_preds[ntype] = (pred, pred_nids)
+
+            save_node_prediction_results(shuffled_preds, save_prediction_path)
+
+        barrier()
+
+
+def node_mini_batch_gnn_predict(model, loader, return_proba=True,
+                                return_label=False, use_ar=False, predict_step=-1):
```

Comment:
Please add a doc string for this function.

Reply:
Added.